Docs/api-gw-jwts-openshift-1-17-x (#19035)
hashicorp/consul: website/content/docs/connect/gateways/api-gateway/configuration/gatewaypolicy.mdx

---
layout: docs
page_title: GatewayPolicy configuration reference
description: Learn about the configuration options for the GatewayPolicy configuration resource. GatewayPolicy resources define API gateway polcies for Consul service mesh on Kubernetes.
---

# GatewayPolicy

> This topic provides reference information for the API gateway policy resource, which defines API gateway policies for Consul service mesh on Kubernetes-orchestrated networks.

Reviewer: should we also include something about how the policy can _only_ target listeners on a gateway and not the entire gateway itself?

> This topic provides reference information for the API gateway policy resource, which defines API gateway policies for Consul service mesh on Kubernetes-orchestrated networks.

Author: ```suggestion
This topic provides reference information for the API gateway policy resource. These resources define policies for listeners on API gateways deployed to Kubernetes-orchestrated networks.
```

## Configuration model

The following list outlines field hierarchy, data types, and requirements in a gateway policy resource. Click on a property name to view additional details, including default values.

- [`apiVersion`](#apiversion): string | required | must be set to `consul.hashicorp.com/v1alpha1`
- [`kind`](#kind): string | required | must be set to `GatewayPolicy`
- [`metadata`](#metadata): map | required
   - [`name`](#metadata-name): string | required 
   - [`namespace`](#metadata-namespace): string  | `default`
- [`spec`](#spec): map | required

>    - [`targetRef`](#spec-targetref): map | required 

Reviewer: this should also have a `namespace` field for the namespace that the target exists in

Author: Just to be clear, we need a `spec.targeRef.namespace` field, yes?

      - [`name`](#spec-targetref): string | required 
      - [`kind`](#spec-targetref): string | required | must be set to `Gateway`

>       - [`group`](#spec-targetref): string | required | must be set to `gateway.networking.kuberenetes.io`

Reviewer: this should actually be `gateway.networking.k8s.io` (or can be said that it needs to match the group of the gateway that it's attaching to)

>       - [`group`](#spec-targetref): string | required | must be set to `gateway.networking.kuberenetes.io`

Author: ```suggestion
      - [`group`](#spec-targetref): string | required 
```
If it can be a different resource group, then we shouldn't say that any specific string is required.

         - [`providers`](#spec-override-providers): list | required 
            - [`name`](#spec-override-providers): string | required 
            - [`verifyClaims`](#spec-override-providers): map | required 
               - [`path`](#spec-override-providers): list of strings | required 
               - [`value`](#spec-override-providers): string | required 
   - [`default`](#spec-default): map | required 
      - [`jwt`](#spec-default-jwt): map | required 
         - [`providers`](#spec-default-providers): list | required 
            - [`name`](#spec-default-providers): string | required 
            - [`verifyClaims`](#spec-default-providers): map | required 
               - [`path`](#spec-default-providers): list of strings | required 
               - [`value`](#spec-default-providers): string | required 

## Complete configuration

When every field is defined, a gateway policy has the following form: 

```yaml
apiVersion: consul.hashicorp.com/v1alpha1
kind: GatewayPolicy
metadata:
   name: <name for the policy>
   namespace: <Consul namespace the policy applies to>
spec:
  targetRef:
    name: gateway
    kind: Gateway
    group: gateway.networking.kuberenetes.io

>     sectionName: <name of a specific section the policy applies>

Reviewer: can we change this to:
```
<name of a specific listener the policy applies>
```

>     sectionName: <name of a specific section the policy applies>

Author: ```suggestion
    sectionName: <name of a specific listener the policy applies>
```

| ---       | ---         | ---       | ---     |
| `name` | Specifies the name of the API gateway that the policy attaches to. | String | None |
| `kind` | Specifies the type of resource that the policy attaches to. Must be set to `Gateway`. | String | None |
| `group` | Specifies the resource group. Must be set to `gateway.networking.kuberenetes.io`. | String | None |
| `sectionName` | Specifies a part of the gateway that the policy applies to. | String | None |

### `spec.override`

Map that contains configurations to apply to listeners when the policy is attached to the gateway. The override settings have precedence over the [`spec.default` configurations](#spec-default).

#### Values

- Default: None
- Data type: Map

### `spec.override.jwt`

Map that contains JWT verification configurations to apply to listeners when the policy is attached to the gateway. The override settings have precedence over the [`default` configurations](#listeners-default) as well as any route-specific JWT configurations defined in [`HTTPRouteAuthFilter` configurations](/consul/docs/connect/gateways/api-gateway/configuration/httprouteauthfilter).

#### Values

- Default: None
- Data type: Map

> ### `spec.override.jwt.providers`

Reviewer: should this specify that it needs to reference an existing JWT provider config entry?

Author: I think we cover that in the usage part of the documentation. I'll add a link to the places you've called out.

               - [`path`](#spec-override-providers): list of strings | required 
               - [`value`](#spec-override-providers): string | required 
   - [`default`](#spec-default): map | required 
      - [`jwt`](#spec-default-jwt): map | required 
         - [`providers`](#spec-default-providers): list | required 
            - [`name`](#spec-default-providers): string | required 
            - [`verifyClaims`](#spec-default-providers): map | required 
               - [`path`](#spec-default-providers): list of strings | required 
               - [`value`](#spec-default-providers): string | required 

## Complete configuration

When every field is defined, a gateway policy has the following form: 

```yaml
apiVersion: consul.hashicorp.com/v1alpha1
kind: GatewayPolicy
metadata:
   name: <name for the policy>
   namespace: <Consul namespace the policy applies to>
spec:
  targetRef:
    name: gateway
    kind: Gateway
    group: gateway.networking.kuberenetes.io
    sectionName: <name of a specific section the policy applies>
  override:
    jwt:
      providers:
      - name: "<name of the provider all listeners verify against>"
        verifyClaims:
        - path:

>           - "<>"

Reviewer 2: I think this should be `"aud"` based on your example above

Author: I want to use a general description in this field similar to the values in the `name` fields, but I don't know what that text is. Using `aud` seems too specific

Author: Ah, OK, thanks. I think I understand. The `api.apps . . .` value is the expected value of an `aud` claim, so it should match that type. The value `user` matches the claim types in the path in the default configuration in this snippet.

#### Values

- Default: None
- This field is required.
- Data type: String

### `metadata.namespace` 

Specifies the namespace that the configuration applies to. Refer to [namespaces](/consul/docs/enterprise/namespaces) for more information.

#### Values

- Default: None
- Data type: String

### `spec`

Map that contains the details about the gateway policy. The `apiVersion`, `kind`, and `metadata` fields are siblings of the `spec` field. All other configurations are children.

#### Values

- Default: None
- This field is required.
- Data type: Map

### `targetRef`

Map that contains references to the gateway that the policy applies to. 

#### Values

- Default: None
- This field is required.
- Data type: Map

The following table describes the members of the `targetRef` map:

| Parameter | Description | Data type | Default |
| ---       | ---         | ---       | ---     |
| `name` | Specifies the name of the API gateway that the policy attaches to. | String | None |
| `kind` | Specifies the type of resource that the policy attaches to. Must be set to `Gateway`. | String | None |

> | `group` | Must be set to `gateway.networking.kuberenetes.io` <TODO need help providing a more complete description/>| String | None |

Author: @nathancoleman @t-eckert - Please help

#### Values

- Default: None
- Data type: Map

### `spec.override.jwt`

Map that contains JWT verification configurations to apply to listeners when the policy is attached to the gateway. The override settings have  precedence over the [`default` configurations](#listeners-default), as well as any route-specific JWT configurations defined in [`HTTPRouteAuthFilter` configurations](/consul/docs/connect/gateways/api-gateway/configuration/httprouteauthfilter).

#### Values

- Default: None
- Data type: Map

### `spec.override.jwt.providers`

Specifies a list of JWT provider configurations to apply to listeners when the policy is attached to the gateway. A provider configuration contains the name of the provider and claims. The override settings have precedence over defaults as well as any listener specific configuration. 

#### Values

- Default: None
- Data type: List of maps

The following table describes the parameters you can specify in a member of the `Providers` list:

| Parameter | Description | Data type | Default |
| ---       | ---         | ---       | ---     |
| `name` | Specifies the name of the provider. | String | None |

> | `verifyClaim` | Specifies a list of paths and a value that define the claim. Consul verifies requests that match the claims declared in the listener JWT configuration and allow the request through the gateway. The `VerifyClaims` map specifies the following settings: <ul><li>`path`: List of paths to the claims</li><li>`value`: ???</li></ul> | Map | None |

Reviewer 2: ```suggestion
| `verifyClaims` | Specifies a list of paths and a value that define the claim. Consul verifies requests that match the claims declared in the listener JWT configuration and allow the request through the gateway. The `VerifyClaims` map specifies the following settings: <ul><li>`path`: List of paths to the claims</li><li>`value`: ???</li></ul> | Map | None |
```

#### Values

- Default: None
- Data type: Map

### `spec.default.jwt`

Specifies default JWT configurations to apply to listeners when the policy is attached to the gateway. You can specify override configurations that have precedence over default configurations. Refer to the [`spec.override.jwt`](#spec-override-jwt) for details.

#### Values

- Default: None
- Data type: Map

### `spec.default.jwt.providers`

Specifies default JWT provider configurations to apply to the listener when the policy is attached to the gateway. A provider configuration contains the name of the provider and claims. You can specify override configurations that have precedence over default configurations. Refer to the [`spec.override.jwt.providers`](#spec-override-jwt-providers) for details.

#### Values

- Default: None
- Data type: List of maps

The following table describes the parameters you can specify in a member of the `Providers` list:

| Parameter | Description | Data type | Default |
| ---       | ---         | ---       | ---     |
| `name` | Specifies the name of the provider. | String | None |

> | `verifyClaim` | Specifies a list of paths and a value that define the claim. Consul verifies requests that match the claims declared in the listener JWT configuration and allow the request through the gateway. The `VerifyClaims` map specifies the following settings: <ul><li>`path`: List of paths to the claims</li><li>`value`: ???</li></ul> | Map | None |

Reviewer 2: ```suggestion
| `verifyClaims` | Specifies a list of paths and a value that define the claim. Consul verifies requests that match the claims declared in the listener JWT configuration and allow the request through the gateway. The `VerifyClaims` map specifies the following settings: <ul><li>`path`: List of paths to the claims</li><li>`value`: ???</li></ul> | Map | None |
```

   - [`default`](#spec-default): map | required 
      - [`jwt`](#spec-default-jwt): map | required 
         - [`providers`](#spec-default-providers): list | required 
            - [`name`](#spec-default-providers): string | required 
            - [`verifyClaims`](#spec-default-providers): map | required 
               - [`path`](#spec-default-providers): list of strings | required 
               - [`value`](#spec-default-providers): string | required 

## Complete configuration

When every field is defined, a gateway policy has the following form: 

```yaml
apiVersion: consul.hashicorp.com/v1alpha1
kind: GatewayPolicy
metadata:
   name: <name for the policy>
   namespace: <Consul namespace the policy applies to>
spec:
  targetRef:
    name: gateway
    kind: Gateway
    group: gateway.networking.kuberenetes.io
    sectionName: <name of a specific listener the policy applies>
  override:
    jwt:
      providers:
      - name: "<name of the provider all listeners verify against>"
        verifyClaims:
        - path:
          - "aud"
          value: "api.apps.organization.com"
  default:
    jwt:
      providers:

>       - name: "<name of the default provider>"

Reviewer: so this provider can be the same as the one in the `override` it's more that these just need to reference by name the specific JWT provider that they should be verified against

>       - name: "<name of the default provider>"

Author: ```suggestion
      - name: "<name of the JWT provider to verify against>"
```

#### Values

- Default: None
- Data type: String

### `spec`

Map that contains the details about the gateway policy. The `apiVersion`, `kind`, and `metadata` fields are siblings of the `spec` field. All other configurations are children.

#### Values

- Default: None
- This field is required.
- Data type: Map

### `targetRef`

Map that contains references to the gateway that the policy applies to. 

#### Values

- Default: None
- This field is required.
- Data type: Map

The following table describes the members of the `targetRef` map:

| Parameter | Description | Data type | Default |
| ---       | ---         | ---       | ---     |
| `namespace` | Specifies the namespace that the target reference is a member of. | String | `default` |
| `name` | Specifies the name of the API gateway that the policy attaches to. | String | None |
| `kind` | Specifies the type of resource that the policy attaches to. Must be set to `Gateway`. | String | None |
| `group` | Specifies the resource group. Must be set to `gateway.networking.kuberenetes.io`. | String | None |
| `sectionName` | Specifies a part of the gateway that the policy applies to. | String | None |

### `spec.override`

> Map that contains configurations to apply to listeners when the policy is attached to the gateway. The override settings have precedence over the [`spec.default` configurations](#spec-default).

Reviewer: should this read `configurations to apply to a listener` rather than `configurations to apply to listeners` since a policy can only apply to a single listener?

#### Values

- Default: None
- Data type: Map

### `spec.override.jwt`

Map that contains JWT verification configurations to apply to listeners when the policy is attached to the gateway. The override settings have precedence over the [`default` configurations](#listeners-default) as well as any route-specific JWT configurations defined in [`RouteAuthFilter` configurations](/consul/docs/connect/gateways/api-gateway/configuration/routeauthfilter).

#### Values

- Default: None
- Data type: Map

### `spec.override.jwt.providers`

> Specifies a list of JWT provider configurations to apply to listeners when the policy is attached to the gateway. A provider configuration contains the name of the provider and claims. The override settings have precedence over defaults as well as any listener specific configuration. Refer to [Use JWTs to verify requests to API gateways on Kubernetes](/consul/docs/connect/gateways/api-gateway/secure-traffic/verify-jwts-k8s) for additional information.

Reviewer: ```suggestion
Specifies a list of JWT provider configurations to apply to listeners when the policy is attached to the gateway listener. A provider configuration contains the name of the provider and claims. The override settings have precedence over defaults as well as any listener specific configuration. Refer to [Use JWTs to verify requests to API gateways on Kubernetes](/consul/docs/connect/gateways/api-gateway/secure-traffic/verify-jwts-k8s) for additional information.
```
